Article: translation of https://geotribu.fr/articles/2025/2025-02-27_travailler-avec-JSON-et-PostgreSQL/ (#34)
geotribu/english-blog: content/posts/2025-01-15_working-with-json-and-postgres.md

---
title: Working with JSON and PostgreSQL
subtitle: Jason and the elephants
authors:

>     - Thomas SZCZUREK-GAYANT

Author: ```suggestion
    - tszczurek
```

categories:
    - article
comments: true

> date: 2025-01-21

Reviewer: Change date. Let's say 2025-02-27? What do you think?

Author: fine with me

> date: 2025-01-21

Author: ```suggestion
date: 2025-02-27
```

description: "Store data in json format in PostgreSQL, consult it... and all this using INSEE french census data as an example. "
icon: simple/postgresql
image:
license: default
robots: index, follow
tags:
    - PostgreSQL
    - JSON
---

> # Working with JSON in PostgreSQL
> 
> :calendar: Initial publication date : {{ page.meta.date | date_localized }}
> 

Reviewer: On this blog, it's automatically set on the left sidebar.

```suggestion
# Working with JSON in PostgreSQL

```

![logo JSON](https://cdn.geotribu.fr/img/logos-icones/programmation/json.png){: .img-thumbnail-left }

As part of a personal project, I wanted to store a large part of the INSEE's french census data in a PostgreSQL database with multi-millennial tables. The problem is that, within the same dataset, the fields can change over the years, which makes it impossible to create a fixed table structure. The solution? Use semi-structured data, i.e. store this data in JSON in a table field. This article is a summary of that experience.

!!! “Unscheduled obsolescence” info
    This work was carried out before the release of PostgreSQL 17, which adds important features for JSON with [`JSON_TABLE`](https://doc.postgresql.fr/17/functions-json.html#FUNCTIONS-SQLJSON-TABLE), so it won't be mentioned here.

Since we're going to be talking about JSON and semi-structured data, I feel obliged to start this article with a warning.

**The relational model is good, eat it up, and integrity constraints were invented for good reason.

> This article is not intended to be an invitation to go into YOLO mode on data management: “all you have to do is put everything in JSON” (like a vulgar dev who would put everything in MongoDB, as the bad tongues would say).
> 

Reviewer: Add the excerpt (https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#adding-an-excerpt):

```suggestion
This article is not intended to be an invitation to go into YOLO mode on data management: “all you have to do is put everything in JSON” (like a vulgar dev who would put everything in MongoDB, as the bad tongues would say).

<!-- more -->
```

> ## JSON for beginners⸱es

Reviewer: Not sure about the median point in English :thinking:

For those of you unfamiliar with [JSON](https://www.json.org/json-fr.html) it's a text-based data representation format from JavaScript that works in part on a `key:value` system that can be seen as a sort of evolution of XML.

```json
{“key_1”: “value”, “key_2”: “value”, “key_3”: “value”}
```

No need for quotation marks for numbers:

```json
{“nb_mushrooms”: 42, “nb_tomatoes”: 31, “first_name”: “roger”}
```

Values can take two forms:

- either a single value, as in the example above,
- or an array, a list, enclosed in `[]`, both of which can be combined in a single JSON object.

```json
{“prenoms”: [“first_name”, “roger”, “fatima”], “nb_mushrooms”: 42}
```

What we call an object is everything between the `{}` used to declare it. To make things even more complex, we can nest objects and give you an example that's a little more meaningful than talking about tomatoes and mushrooms:

> ```json title="Exemple de JSON des superhéros" linenums="1"

Reviewer: Code-block title is untranslated

> ```json title="Exemple de JSON des superhéros" linenums="1"

Author: ```suggestion
```json title="Example with JSON of superheroes" linenums="1"
```

### With nested JSON

Well, he's a nice guy, but his JSON is still JSON where everything is on the first level. Well, for my needs, that was enough. But I'm not going to run away from my responsibilities and we'll see how it goes with more complex JSON.  
I'll use the example from the previous section to make things more complex after this aside.

To inject complex json into a field, we have two solutions: nesting dedicated functions, or castering a text string. Let's imagine a “test” table, in a “test” schema, with a `jsonb` field named “donnees”.

```sql
INSERT INTO test.test (donnees) VALUES
(jsonb_object(
 'key1', 'value1',
 'key2', jsonb_array(
  'foo', 'bar', 'baz')
  )
 )
```

Cette insertion pourrait tout aussi bien s'écrire avec un cast d'une chaine de texte vers du jsonb. Attention, la syntaxe json doit être ici respectée :

```sql
INSERT INTO test.test (donnees) VALUES
('{"key1": "value1", "key2": ["foo", "bar", "baz"]}'::jsonb)
```

> #### Recupération

Reviewer: Untranslated

> #### Recupération

Author: ```suggestion
#### Data recovery
```

   pop_25_54_ans_employes_c,' || POP2554_CS5_C || ',
   pop_25_54_ans_ouvriers_c,' || POP2554_CS6_C || ',
   pop_25_54_ans_retraites_c,' || POP2554_CS7_C || ',
   pop_25_54_ans_autres_c,' || POP554_CS8_C || ',
   population_55_ans_et_plus_c,' || POP55P_C || ',
   pop_55_ans_et_plus_ans_agriculteurs_c,' || POP55P_CS1_C || ',
   pop_55_ans_et_plus_ans_artisants_commercants_chefs_entreprise_c,' || POP55P_CS2_C || ',
   pop_55_ans_et_plus_ans_cadres_prof_intel_sup_c,' || POP55P_CS3_C || ',
   pop_55_ans_et_plus_ans_professions_intermediaires_c,' || POP55P_CS4_C || ',
   pop_55_ans_et_plus_ans_employes_c,' || POP55P_CS5_C || ',
   pop_55_ans_et_plus_ans_ouvriers_c,' || POP55P_CS6_C || ',
   pop_55_ans_et_plus_ans_retraites_c,' || POP55P_CS7_C || ',
   pop_55_ans_et_plus_ans_autres_c,' || POP55P_CS8_C,
  E'[\t\n\r]','','g') AS data
  FROM insee.rp_population_import
 )

-- Conversion of text strings to json and insertion into the table.
INSERT INTO insee.donnees_communes("code_commune","annee","fk_base","donnees")
SELECT
    CODGEO,
    2021,
    1,
    jsonb_object(string_to_array(data::text,','))
FROM d
ORDER BY "CODGEO";
```

> Ouf.

Reviewer: ```suggestion
_Ouf._
```

---
title: Working with JSON and PostgreSQL
subtitle: Jason and the elephants
authors:
    - tszczurek
categories:
    - article
comments: true
date: 2025-02-27
description: "Store data in json format in PostgreSQL, consult it... and all this using INSEE french census data as an example. "
icon: simple/postgresql
image:
license: default
links:
  - Original version (French): https://geotribu.fr/articles/2025/2025-01-21_travailler-avec-JSON-et-PostgreSQL/
robots: index, follow
tags:
    - PostgreSQL
    - JSON
---

# Working with JSON in PostgreSQL

![logo JSON](https://cdn.geotribu.fr/img/logos-icones/programmation/json.png){: .img-thumbnail-left }

As part of a personal project, I wanted to store a large part of the INSEE's french census data in a PostgreSQL database with multi-millennial tables. The problem is that, within the same dataset, the fields can change over the years, which makes it impossible to create a fixed table structure. The solution? Use semi-structured data, i.e. store this data in JSON in a table field. This article is a summary of that experience.

> !!! “Unscheduled obsolescence” info

Reviewer: Classic AI bot artifact spotted :wink: 

```suggestion
!!! info "Unscheduled obsolescence"
```

    This work was carried out before the release of PostgreSQL 17, which adds important features for JSON with [`JSON_TABLE`](https://doc.postgresql.fr/17/functions-json.html#FUNCTIONS-SQLJSON-TABLE), so it won't be mentioned here.

Since we're going to be talking about JSON and semi-structured data, I feel obliged to start this article with a warning.

**The relational model is good, eat it up, and integrity constraints were invented for good reason.**

This article is not intended to be an invitation to go into YOLO mode on data management: “all you have to do is put everything in JSON” (like a vulgar dev who would put everything in MongoDB, as the bad tongues would say).

<!-- more -->
## JSON for beginners

For those of you unfamiliar with [JSON](https://www.json.org/json-fr.html) it's a text-based data representation format from JavaScript that works in part on a `key:value` system that can be seen as a sort of evolution of XML.

```json
{“key_1”: “value”, “key_2”: “value”, “key_3”: “value”}
```

No need for quotation marks for numbers:

```json

> {“nb_mushrooms”: 42, “nb_tomatoes”: 31, “first_name”: “roger”}

Author: ```suggestion
{"nb_mushrooms": 42, "nb_tomatoes": 31, "first_name": "roger"}
```

```

Values can take two forms:

- either a single value, as in the example above,
- or an array, a list, enclosed in `[]`, both of which can be combined in a single JSON object.

```json

> {“prenoms”: [“first_name”, “roger”, “fatima”], “nb_mushrooms”: 42}

Author: ```suggestion
{"prenoms": ["first_name", "roger", "fatima"], "nb_mushrooms": 42}
```

> {"prenoms": ["first_name", "roger", "fatima"], "nb_mushrooms": 42}

Author: ```suggestion
{"prenoms": ["elodie", "roger", "fatima"], "nb_mushrooms": 42}
```

      "age": 1000000,
      "secretIdentity": "Unknown",
      "powers": [
        "Immortality",
        "Heat Immunity",
        "Inferno",
        "Teleportation",
        "Interdimensional travel"
      ]
    }
  ]
}
```

An example from the [Mozilla](https://developer.mozilla.org/fr/docs/Learn/JavaScript/Objects/JSON) site, which will give you a better idea. You can also consult the [wikipedia](https://fr.wikipedia.org/wiki/JavaScript_Object_Notation) article or [the infamous specification site](https://www.json.org/json-en.html).

## PostgreSQL's json types

PostgreSQL is able to store data/objects in json format in fields that are assigned a dedicated type. There are two types, because otherwise it wouldn't be any fun.

- The `json` type is there for historical reasons, to allow databases that used this type in the past to function. It stores information in text form, which is not optimized for a computer. There is, however, an advantage to using it: it allows you to retrieve information on key order. If it's important for you to know that `name` is key 1 and `firstname` is key 2, without having to go through the key name again, then you'll need to use the `json` type.

> - the `jsonb` type. The modern type. It stores information in binary form and offers a lott of functions in addition to those of the `json` type.

Author: ```suggestion
- the `jsonb` type. The modern type. It stores information in binary form and offers a lot of functions in addition to those of the `json` type.
```

UNIQUE (fk_base, annee, code_commune)
CONSTRAINT fk_donnees_bases FOREIGN KEY fk_base ON insee.bases(pk_id) ON UPDATE CASCADE ON DELETE CASCADE
);

-- Index creation
-- A multi-field index is already automatically created on (fk_base, annee, code_commune) thanks to the UNIQUE clause in the table declaration.
CREATE INDEX idx_donnees_com_donnees ON insee.donnees_communes USING gin (donnees);
END;
```

The most fussy of you DBAs will have noticed this `CHECK`. “Why doesn't he use varchar(5)? It really doesn't make sense!” Quite simply, because this form allows you to use a text type with **really** arbitrary numbers of characters (the `text` type) unlike varchar(255), while being able to control the minimum and maximum number with the `CHECK` predicate (unlike varchar which only controls the maximum) as explained on the [Postgres wiki].(<https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_varchar.28n.29_by_default>).

And we insert a few rows in our insee.bases table :

```sql title="SQL script for data insertion" linenums="1"
INSERT INTO insee.bases (nom) VALUES
('rp_population'),
('rp_menages'),
('rp_logement'),
('rp_diplomes'),
('rp_activite'),
('rp_emploi')
```

> ## Insertion de données et récupération

Author: ```suggestion
## Data insertion and retrieval
```

If you're still here, I assume you already know this. However, the form `(something)::(type)` is a PostgreSQL shortcut for making a `cast`, i.e. converting a value into another type. With `->>` the value is returned to us as text and we convert it into an integer.

### With nested JSON

Well, he's a nice guy, but his JSON is still JSON where everything is on the first level. Well, for my needs, that was enough. But I'm not going to run away from my responsibilities and we'll see how it goes with more complex JSON.  
I'll use the example from the previous section to make things more complex after this aside.

To inject complex json into a field, we have two solutions: nesting dedicated functions, or castering a text string. Let's imagine a “test” table, in a “test” schema, with a `jsonb` field named “donnees”.

```sql
INSERT INTO test.test (donnees) VALUES
(jsonb_object(
 'key1', 'value1',
 'key2', jsonb_array(
  'foo', 'bar', 'baz')
  )
 )
```

> Cette insertion pourrait tout aussi bien s'écrire avec un cast d'une chaine de texte vers du jsonb. Attention, la syntaxe json doit être ici respectée :

Reviewer 2: Mmmmh... :thinking: 

Voulay-vous couchay avec moi, ce swar ??

> Cette insertion pourrait tout aussi bien s'écrire avec un cast d'une chaine de texte vers du jsonb. Attention, la syntaxe json doit être ici respectée :

Author: ```suggestion
This insertion could just as easily be written as a cast from a text string to jsonb. Please note that the json syntax must be respected here:
```

Well, now that we've tried to explain the concepts as best we can with simple examples, because we've got to start somewhere, let's try something a little more voluminous.

Let's retrieve the latest population data from the municipal census [in csv format from the Insee website](https://www.insee.fr/fr/information/8183122) (you'll want the bases for the main indicators, sorry page not available in english). For this example, we'll use the “Evolution et structure de la population” file. First, clean up the field names. Insee systematically indicates the year in its field names, which means that they change every year for the same indicator.

All field names begin with P or C, indicating *main survey* (raw answers to census questions) or *complementary survey* (cross-referencing of answers to establish an indicator). Fields from the main survey and those from the complementary survey must not be cross-referenced. This information should obviously be kept, but for personal reasons, I prefer to put it at the end of the name rather than at the beginning. In this way, we move from normalized fields such as `P18_POP` to a normalization of this type `POP_P`.

> You'll find [here](https://github.com/thomas-szczurek/base_donnees_insee/blob/main/sql/import/correction_champs_insee.xlsx) a spreadsheet to take care of all this.

Author: ```suggestion
You'll find [here](https://gitlab.com/Thomas_szczurek_gayant/bd_insee/-/blob/main/sql/import/correction_champs_insee.xlsx?ref_type=heads) a spreadsheet to take care of all this.
```

Before inserting the data into our table, we'll go through a temporary table to make the data accessible in Postgres. Using Postgresql's `COPY` would be tedious, as you'd have to specify the hundred or so fields contained in the census population section of the command. And I'm not ashamed to say that I've got a baobab in my hand at the thought. So we pull out this wonderful software called QGIS. We activate the Explorer and Explorer2 panels. We create a connection to the database with creation rights, and with a graceful flick of the wrist, we drag the file from the Explorer panel to the Postgres database in Explorer2. Let the magic happen.

Now, get ready for perhaps one of the weirdest INSERTs of your life (at least, it was for me!). Ugh. I realize that if I wanted to do this right, I'd also have to explain [CTE](https://www.postgresql.org/docs/current/queries-with.html). But I'll let you click on the link so as not to make it too long.

> We're going to use the CTE to concatenate the name we want to give to our keys with the values contained in our temporary table in a string separated by `,`. We'll send it to a `string_to_array()` function and then to a `jsonb_object()` function. We'll also take the opportunity to clean up any tabs or carriage returns that might remain with a regular expression, using the `regex_replace()` function. (These characters are called ``t`,``n`and ``r`). The latter function takes 3 arguments: the source string, the searched-for `pattern` and the replacement text. The optional `g` flag* is added to replace all occurrences found.

Author: ```suggestion
We're going to use the CTE to concatenate the name we want to give to our keys with the values contained in our temporary table in a string separated by `,`. We'll send it to a `string_to_array()` function and then to a `jsonb_object()` function. We'll also take the opportunity to clean up any tabs or carriage returns that might remain with a regular expression, using the `regex_replace()` function. (These characters are called `\t`,`\n`and `\r`). The latter function takes 3 arguments: the source string, the searched-for `pattern` and the replacement text. The optional `g` flag* is added to replace all occurrences found.
```

   pop_25_54_ans_employes_c,' || POP2554_CS5_C || ',
   pop_25_54_ans_ouvriers_c,' || POP2554_CS6_C || ',
   pop_25_54_ans_retraites_c,' || POP2554_CS7_C || ',
   pop_25_54_ans_autres_c,' || POP554_CS8_C || ',
   population_55_ans_et_plus_c,' || POP55P_C || ',
   pop_55_ans_et_plus_ans_agriculteurs_c,' || POP55P_CS1_C || ',
   pop_55_ans_et_plus_ans_artisants_commercants_chefs_entreprise_c,' || POP55P_CS2_C || ',
   pop_55_ans_et_plus_ans_cadres_prof_intel_sup_c,' || POP55P_CS3_C || ',
   pop_55_ans_et_plus_ans_professions_intermediaires_c,' || POP55P_CS4_C || ',
   pop_55_ans_et_plus_ans_employes_c,' || POP55P_CS5_C || ',
   pop_55_ans_et_plus_ans_ouvriers_c,' || POP55P_CS6_C || ',
   pop_55_ans_et_plus_ans_retraites_c,' || POP55P_CS7_C || ',
   pop_55_ans_et_plus_ans_autres_c,' || POP55P_CS8_C,
  E'[\t\n\r]','','g') AS data
  FROM insee.rp_population_import
 )

-- Conversion of text strings to json and insertion into the table.
INSERT INTO insee.donnees_communes("code_commune","annee","fk_base","donnees")
SELECT
    CODGEO,
    2021,
    1,
    jsonb_object(string_to_array(data::text,','))
FROM d
ORDER BY "CODGEO";
```

> *Ouf.*

Author: ```suggestion
*Wow.*
```

Be sure to group the inserts in this order: year/base/commune_code, to make it easier for PostgreSQL to read the data.

Now, imagine that, like the author of these lines, your fat, pudgy fingers stumble and slip on your keyboard while writing this query, that a typo slips in, and then it's drama. How do you change the name of a key already encoded in the table? Here's how:

```sql
CREATE TABLE example(id int PRIMARY KEY, champ jsonb);
INSERT INTO example VALUES
    (1, '{"nme": "test"}'),
    (2, '{"nme": "second test"}');

UPDATE example
SET champ = champ - 'nme' || jsonb_build_object('name', champ -> 'nme')
WHERE champ ? 'nme'
returning *;
```

`-` is an operator used to remove a key from a json object. For our UPDATE, we therefore remove our typo from the entire object. In addition, we concatenate everything else with the construction of a new object in which we correct the key name. We also assign the value of the key being deleted, which is still usable at UPDATE time, to the fields that originally contained it with `?`.

> ## Now what? What do we do with this?

Reviewer 2: ```suggestion
## Now what ? What do we do with this ?
```

ORDER BY fk_base, clef_json;
```

![Overview of vm with the first and last year of key presence](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/vm_presence_cles_annees.png){: .img-center loading=lazy }

The only difficulty here is the presence of [LATERAL](https://www.postgresql.org/docs/current/queries-table-expressions.html#QUERIES-FROM). This “word” allows you to use a field from the main query in a subquery placed in a FROM clause. The elements of the subquery will then be evaluated atomically before being joined to the original table. Yes, it's not very easy to explain. Here, the WHERE of the subquery will query the `donnees` field of the “donnees_communes” table to see if it sees the clef_json (json key) being evaluated in the “correspondance_clefs_champs” table with alias “c”. If so, we take the minimum/maximum value of the year field and attach it to this line and this line only. Then evaluate the following clef_json ... (*evaluated atomically*)

Now we'd like to offer a product that's a little easier to use with all this. Excuse me in advance, I'll copy a 50-line query a second time.

> The only table used here that we haven't created is a `zonages_administratifs` table containing municipalities codes in a `code_admin` field and a `fk_type` field containing the type of administrative zoning (1 for communes).

Author: ```suggestion
The only table used here that we haven't created is a `zonages_administratifs` table containing municipalities codes in a `code_admin` field and a `fk_type` field containing the type of administrative zoning (1 for municipalities).
```
